WordPress/Classes/Services/PostService.m

```diff
@@ -333,70 +337,129 @@ - (void)autoSave:(AbstractPost *)post
-                        success(nil, nil);
+                        success(postInContext, previewURL);
                     }
+                }];
+            } else {
+                // This can happen if the post was deleted right after triggering the auto-save.
+                if (success) {
+                    success(nil, nil);
                 }
-            }];
-        };
-        
-        // The autoSave endpoint returns an exception on posts that do not exist on the server
-        // so we'll create the post instead if necessary.
-        BOOL mustBeCreated = ![post hasRemote];
-
-        if (mustBeCreated) {
-            // Abort if the status is trashed/deleted. We'd rather not automatically create a
-            // locally trashed post as drafts in the server.
-            if ([post.status isEqualToString:PostStatusTrash] || [post.status isEqualToString:PostStatusDeleted]) {
-                setPostAsFailedAndCallFailureBlock(defaultError);
-                return;
             }
+        }];
+    };
+
+}
+
+- (void)autoSave:(AbstractPost *)post
+         success:(void (^)(AbstractPost *post, NSString *previewURL))success
+         failure:(void (^)(NSError * _Nullable error))failure
+{
+    id<PostServiceRemote> remote = [self.postServiceRemoteFactory forBlog:post.blog];
+    if ([remote isKindOfClass:[PostServiceRemoteREST class]]) {
+        [self handleAutoSaveWithRestRemote:(PostServiceRemoteREST *)remote forPost:post success:success failure:failure];
 
-            [self uploadPost:post
-        forceDraftIfCreating:YES
-                     success:^(AbstractPost * _Nonnull post) {
-                         success(post, nil);
-                     }
-                     failure:failure];
-        } else {
-            [restRemote autoSave:remotePost
-                         success:successBlock
-                         failure:setPostAsFailedAndCallFailureBlock];
-        }
     } else if ([post originalIsDraft] && [post isDraft]) {
         [self uploadPost:post
                  success:^(AbstractPost * _Nonnull post) {
                      success(post, nil);
                  } failure:failure];
+
     } else {
-        setPostAsFailedAndCallFailureBlock(defaultError);
+        NSError *defaultError = [self defaultAutosaveError];
+        AutosaveFailureBlock failureBlock = [self wrappedAutosaveFailureBlock:post failure:failure];
+        failureBlock(defaultError);
     }
 }
 
+- (void)handleAutoSaveWithRestRemote:(PostServiceRemoteREST *)restRemote
+                             forPost:(AbstractPost *)post
+                             success:(void (^)(AbstractPost *post, NSString *previewURL))success
+                             failure:(void (^)(NSError * _Nullable error))failure
+{
+    NSManagedObjectID *postObjectID = post.objectID;
+    RemotePost *remotePost = [self remotePostWithPost:post];
+
+    AutosaveSuccessBlock successBlock = [self wrappedAutosaveSuccessBlock:postObjectID success:success];
```

Comment:
In line with my comment below about the `failureBlock` name, we could also rename this to `autosaveSuccessBlock`. It's not a great name but it'll probably make it less surprising why we use `success` and `successBlock` sometimes. 

Admittedly, I didn't see this ambiguous stuff when I last touched this method. 😄